gneiss/plot/heatmap.py

```diff
@@ -0,0 +1,124 @@
+
+Functions
+---------
+
+.. autosummary::
+   :toctree: generated/
+
+   heatmap
+
+"""
+
+import numpy as np
+import pandas as pd
+from skbio.stats.composition import clr_inv
+from collections import OrderedDict
+from ete3 import Tree, TreeStyle, AttrFace, ProfileFace
+from ete3 import ClusterNode
+from ete3.treeview.faces import add_face_to_node
+from gneiss.layouts import default_layout
+from gneiss.balances import _attach_balances
+import io
+
+
+def heatmap(table, tree, layout=None, cmap='viridis', **kwargs):
+    """ Plots tree on heatmap
+
+    Parameters
+    ----------
+    table : pd.DataFrame
+        Contingency table where samples correspond to rows and
+        features correspond to columns.
+    tree : skbio.TreeNode
+        A strictly bifurcating tree defining a hierarchical relationship
+        between all of the features within `table`.
+    layout : function, optional
+        A layout for formatting the tree visualization. Must take a
+        `ete.tree` as a parameter.
+    cmap: matplotlib colormap
+        String or function encoding matplotlib colormap.
+    rowlabel_size : int
+        Size of row labels
+    width : int
+        Heatmap cell width
+    height : int
+        Heatmap cell height
+
+    Returns
+    -------
+    ete.Tree
+        ETE tree object that will be plotted.
+    ete.TreeStyle
+        ETE TreeStyle that decorates the tree and heatmap visualization.
+    """
+
+    params = {'rowlabel_size': 8, 'width':200, 'height':14, 'cmap':'viridis'}
+    for key in params.keys():
+        params[key] = kwargs.get(key, params[key])
+    fsize = params['rowlabel_size']
+    width = params['width']
+    height = params['height']
+    colorscheme = params['cmap']
+
+
+    # Allow for matplotlib colors to be encoded in ETE3 heatmap
+    # Originally from https://github.com/lthiberiol/virfac
+    def get_color_gradient(self):
+        from PyQt4 import QtGui
+        try:
+            import matplotlib.pyplot as plt
+            import matplotlib.colors as colors
+            import matplotlib.cm as cmx
+        except:
+            ImportError("Matplotlib not installed.")
+
+        cNorm  = colors.Normalize(vmin=0, vmax=1)
+        scalarMap = cmx.ScalarMappable(norm=cNorm, cmap=plt.get_cmap(self.colorscheme))
+        color_scale = []
+        for scale in np.linspace(0, 1, 255):
+            [r, g, b, a] = scalarMap.to_rgba(scale, bytes=True)
+            color_scale.append( QtGui.QColor( r, g, b, a ) )
+        return color_scale
+
+    ProfileFace.get_color_gradient = get_color_gradient
+    tree.name = ""
+
+    f = io.StringIO()
+    table.T.to_csv(f, sep='\t', index_label='#Names')
+
+    tr = ClusterNode(str(tree), text_array=str(f.getvalue()))
+    matrix_max = np.max(table.values)
+    matrix_min = np.min(table.values)
+    matrix_avg = matrix_min + ((matrix_max -matrix_min) / 2)
+
+    # Encode the actual profile face
+    nameFace = AttrFace("name", fsize=fsize)
+    def heatmap_layout(node):
```

Comment:
That is correct.  We are using the layout function to decorate the leaves of the tree.